### PR TITLE
Parameter Regression

### PR DESCRIPTION
**Pull request by tanzim10**
### Overview
This pull request introduces a new class designed for parameter regression analysis, specifically tailored to model the movement dynamics of user equipment (UE). The class uses polynomial regression and least squares optimization to estimate the parameter `alpha`, which is crucial for predicting UE movement based on current states.

### Key Features
- **Polynomial Regression**: Employs polynomial regression to fit the model to the observed UE data.
- **Alpha Estimation**: Uses least squares optimization to fine-tune the parameter `alpha`, ensuring minimal residuals.
- **State Dependency Modeling**: Models how future UE states depend on current states, enhancing the predictiveness of the model.

### Technical Details
- The regression model calculates `alpha` to balance the historical impact on predicted states, offering a refined approach to understanding UE dynamics.
- Added robust unit tests to validate the correctness of the regression analysis and the alpha parameter estimation under various simulated conditions.

### Unit Tests
- **Test Accuracy**: Ensures the regression model accurately predicts UE states.
- **Test Robustness**: Checks the model's performance against edge cases and unusual datasets.
- **Parameter Sensitivity**: Assesses how changes in input parameters affect the model's output, ensuring stability and reliability.

### CODE REVIEW

File: radp/digital_twin/mobility/param_regression.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        v_t_next = (
+            alpha * v_t + alpha2 * self.velocity_mean + alpha3 * np.random.normal()
+        )
+        angle_mean = theta_t  # Simplified model without margin correction
+        theta_t_next = (
+            alpha * theta_t + alpha2 * angle_mean + alpha3 * np.random.normal()
+        )
+        return np.array([v_t_next, theta_t_next])
+
+    def residual_vector(
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/param_regression.py:87]: make this also static

File: radp/digital_twin/mobility/param_regression.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
+import numpy as np
+from scipy import optimize
+from typing import Tuple
+
+
+class ParameterRegression:
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/param_regression.py:6]: consider making this a pure API (no class needed)

File: radp/digital_twin/mobility/tests/test_param_regression.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        ------------
+        To run these tests, execute the following command in your terminal:
+        ```
+        python3 -m unittest radp/digital_twin/mobility/tests/test_parameter_regression.py
+        ```
+        """
+        data = {
+            "mock_ue_id": np.random.randint(0, 10, size=100),
+            "velocity": np.random.uniform(0, 100, size=100),
+        }
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/tests/test_param_regression.py:26]: let's make this a fixed dataset -- choose just 2 UEs, with just 5 track points each

File: radp/digital_twin/mobility/tests/test_param_regression.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    def test_optimize_alpha_output(self) -> None:
+        """
+        Test that optimize_alpha returns an optimized alpha and a covariance matrix.
+
+        Asserts that popt is an instance of np.ndarray and pcov is also an instance
+        of np.ndarray. Additionally checks that popt contains a single optimized value.
+        """
+        alpha0 = [0.8]
+        popt, pcov = self.model.optimize_alpha(alpha0)
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/tests/test_param_regression.py:65]: this can be asserted exactly

File: radp/digital_twin/mobility/tests/test_param_regression.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    def test_velocity_mean_calculation(self) -> None:
+        """
+        Test that the velocity mean is calculated correctly.
+
+        Asserts that the calculated mean of the velocity in the model matches the
+        expected mean calculated from the DataFrame.
+        """
+        expected_mean = np.mean(self.df["velocity"].to_numpy())
+        self.assertAlmostEqual(self.model.velocity_mean, expected_mean, places=5)
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/tests/test_param_regression.py:80]: may delete

File: radp/digital_twin/mobility/tests/test_param_regression.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    def test_variance_calculation(self) -> None:
+        """
+        Test that the variance of velocity is calculated correctly.
+
+        Asserts that the calculated variance of the velocity in the model matches the
+        expected variance calculated from the DataFrame.
+        """
+        expected_variance = np.var(self.df["velocity"].to_numpy())
+        self.assertAlmostEqual(self.model.variance, expected_variance, places=5)
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/tests/test_param_regression.py:90]: may delete

File: radp/digital_twin/mobility/tests/test_param_regression.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+        Asserts that the shape of the calculated theta_t_full matches the expected shape
+        after applying the polynomial function and adding noise.
+        """
+        f = np.poly1d([8, 7, 5, 1])
+        expected_theta_t_full = f(self.model.v_t_full) + 6 * np.random.normal(
+            size=len(self.model.v_t_full)
+        )
+
+        self.assertEqual(self.model.theta_t_full.shape, expected_theta_t_full.shape)
```
Comment (paulvarkey) [inline radp/digital_twin/mobility/tests/test_param_regression.py:104]: may delete

**Review by paulvarkey — CHANGES_REQUESTED** [review]
make a separate black PR and then this PR can be focussed on actual changes (with suggested review comments)
